### ISSUE TITLE
baseRequest and request have a null context it GzipHandler

### ISSUE THREAD

**Issue by lmame** [eclipse/jetty.project#2744]
Hi :)

First of all I am a UI developer so take it easy :)

**What is working:**
We coded a platform using Spring / Jax Rx and Jetty 9.3.7.v20160115 osgi to load bundles (it can be libraries or applications with java / js / html / css).
Our java process launches Jetty passing jetty.home which loads a jetty.xml and the other xml files.
So far, so good.

I have been tasked to activate gzip compression, so I used the jetty-gzip.html provided by Jetty and declared the jetty.etc.config.urls to load the new file:
jetty.etc.config.urls=etc/jetty.xml,etc/jetty-http.xml,etc/jetty-deployer.xml,etc/jetty-gzip.xml
The server starts, the resources are correctly gzipped, no problem.

**But:**
Then I tried phase two. Instead of making Jetty compress the file, I could provide some resources pre-gzipped and tell Jetty to use that if they are available (we can use grunt to do that for us when we build the bundles).
I created a .gz myself with a flag inside file different than the non gzipped js to know what was loaded:
test.js (with flag A)
test.js.gz (with flag B)
I set jetty.gzip.checkGzExists to true in the jetty-gzip.xml and restarted the server.
I ask for test,js, Jetty sends me gzipped data (good) but with flag A. Which means that even if the .gz exists, it has not been used.

At this point I got Jetty source code and did a debug remote connection to the server and set a breakpoint in the GzipHandler handle method:
```
public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
// (...)
LOG.debug("{} handle {} in {}",this,baseRequest,context);

```
What I see there is that the context of baseRequest and request are null.
So when a test is done later the if condition fails and no check is done on the real path because context is null:
```
if (_checkGzExists && context!=null)
        {
            String realpath=request.getServletContext().getRealPath(path);
            if (realpath!=null)
            {
                File gz=new File(realpath+".gz");
```
The (strange) thing is that right after in the code the context is set by this call:
```
// install interceptor and handle
        out.setInterceptor(new GzipHttpOutputInterceptor(this,_vary,baseRequest.getHttpChannel(),out.getInterceptor(),_syncFlush));
        if (_handler!=null)
            _handler.handle(target,baseRequest, request, response);

```

That will ultimately call the setContext in Request:
```
    public void setContext(Context context)
    {
        _newContext = _context != context;
        _context = context;
    }
```

So it seems that as least in our case the GZipHandler is called before the context is set.

**Another test:**
I tried to use the latest Jetty (9.4.11.v20180605). I rebuilt our project, started the server and I see the same behavior.
In GzipHandler the context of baseRequest and request are null.

**Anyway...:**
I saw that in Jetty 10 checkGzExists will be removed and not replaced.
Does this mean the static option using a pre-gzipped file will be removed as well or it will be enabled all the time?
Is the behavior I am seeing normal, or I did something totally wrong when setting the GzipHandler?

Fyi in our server (or bundle code) we have some rest apis and the request context is correctly set (not null).

**Comment by joakime** [comment]
Be aware that `jetty.etc.config.urls` are **very order dependent** (also, your list of XMLs is incomplete).

Jetty home (and also Jetty distribution) manages the order based on the feature/module dependency tree, but OSGi cannot handle that.

If you are using the XML files from the jetty-home artifact, or jetty-distribution, or other Eclipse Jetty source, then you'll probably want to at least install the jetty-home artifact in a unique directory and use it to glean the information you want/need.

Eg:

```
$ mkdir issue-2744-base
$ cd issue-2744-base/

[issue-2744-base]$ java -jar ../../jetty-home-9.4.11.v20180605/start.jar --add-to-start=http,deploy,gzip
INFO  : webapp          transitively enabled, ini template available with --add-to-start=webapp
INFO  : server          transitively enabled, ini template available with --add-to-start=server
INFO  : security        transitively enabled
INFO  : servlet         transitively enabled
INFO  : http            initialized in ${jetty.base}/start.ini
INFO  : gzip            initialized in ${jetty.base}/start.ini
INFO  : threadpool      transitively enabled, ini template available with --add-to-start=threadpool
INFO  : deploy          initialized in ${jetty.base}/start.ini
MKDIR : ${jetty.base}/webapps
INFO  : Base directory was modified

[issue-2744-base]$ java -jar ../../jetty-home-9.4.11.v20180605/start.jar --list-config

Java Environment:
-----------------
 java.home = /mnt/c/Users/joakim/jdks/wsl/jdk-9 (null)
 java.vm.vendor = Oracle Corporation (null)
 java.vm.version = 9+181 (null)
 java.vm.name = Java HotSpot(TM) 64-Bit Server VM (null)
 java.vm.info = mixed mode (null)
 java.runtime.name = Java(TM) SE Runtime Environment (null)
 java.runtime.version = 9+181 (null)
 java.io.tmpdir = /tmp (null)
 user.dir = /mnt/c/code/jetty/distros/bases/issue-2744-base (null)
 user.language = en (null)
 user.country = US (null)

Jetty Environment:
-----------------
 jetty.version = 9.4.11.v20180605
 jetty.tag.version = master
 jetty.home = /mnt/c/code/jetty/distros/jetty-home-9.4.11.v20180605
 jetty.base = /mnt/c/code/jetty/distros/bases/issue-2744-base

Config Search Order:
--------------------
 <command-line>
 ${jetty.base} -> /mnt/c/code/jetty/distros/bases/issue-2744-base
 ${jetty.home} -> /mnt/c/code/jetty/distros/jetty-home-9.4.11.v20180605


JVM Arguments:
--------------
 (no jvm args specified)

System Properties:
------------------
 (no system properties specified)

Properties:
-----------
 java.version = 9
 java.version.major = 9
 java.version.micro = 0
 java.version.minor = 0
 java.version.platform = 9
 jetty.base = /mnt/c/code/jetty/distros/bases/issue-2744-base
 jetty.base.uri = file:///mnt/c/code/jetty/distros/bases/issue-2744-base
 jetty.home = /mnt/c/code/jetty/distros/jetty-home-9.4.11.v20180605
 jetty.home.uri = file:///mnt/c/code/jetty/distros/jetty-home-9.4.11.v20180605

Jetty Server Classpath:
-----------------------
Version Information on 11 entries in the classpath.
Note: order presented here is how they would appear on the classpath.
      changes to the --module=name command line options will be reflected here.
 0:                    3.1.0 | ${jetty.home}/lib/servlet-api-3.1.jar
 1:                 3.1.0.M0 | ${jetty.home}/lib/jetty-schemas-3.1.jar
 2:         9.4.11.v20180605 | ${jetty.home}/lib/jetty-http-9.4.11.v20180605.jar
 3:         9.4.11.v20180605 | ${jetty.home}/lib/jetty-server-9.4.11.v20180605.jar
 4:         9.4.11.v20180605 | ${jetty.home}/lib/jetty-xml-9.4.11.v20180605.jar
 5:         9.4.11.v20180605 | ${jetty.home}/lib/jetty-util-9.4.11.v20180605.jar
 6:         9.4.11.v20180605 | ${jetty.home}/lib/jetty-io-9.4.11.v20180605.jar
 7:         9.4.11.v20180605 | ${jetty.home}/lib/jetty-security-9.4.11.v20180605.jar
 8:         9.4.11.v20180605 | ${jetty.home}/lib/jetty-servlet-9.4.11.v20180605.jar
 9:         9.4.11.v20180605 | ${jetty.home}/lib/jetty-webapp-9.4.11.v20180605.jar
10:         9.4.11.v20180605 | ${jetty.home}/lib/jetty-deploy-9.4.11.v20180605.jar

Jetty Active XMLs:
------------------
 ${jetty.home}/etc/jetty-threadpool.xml
 ${jetty.home}/etc/jetty.xml
 ${jetty.home}/etc/jetty-webapp.xml
 ${jetty.home}/etc/jetty-deploy.xml
 ${jetty.home}/etc/jetty-gzip.xml
 ${jetty.home}/etc/jetty-http.xml
```

From here you'll see what the order of XML is.
Seems like you are missing a few.

> test.js (with flag A)
> test.js.gz (with flag B)
> I set jetty.gzip.checkGzExists to true in the jetty-gzip.xml and restarted the server.
> I ask for test,js, Jetty sends me gzipped data (good) but with flag A. Which means that even if the .gz exists, it has not been used.

The `checkGzExists` method is pointless, don't bother with it.
That `test.js` file is being served by the `DefaultServlet`.
Unless you are using Spring or Jax RS, which decides that they don't want Jetty to serve static files, but will do it themselves.
You'll have to turn off all of those features in Jax RS first (a common configuration), then test again.
If it still behaves like this, then you probably also have Spring serving static files (much rarer configuration).

> What I see there is that the context of baseRequest and request are null.
> So when a test is done later the if condition fails and no check is done on the real path because context is null:

`GzipHandler` can operate in a scenario where there is no `ServletContext` (that's quite common in fact), but having a null Requests is downright strange.
Fix your XML setup (from earlier), and try again.

> request.getServletContext().getRealPath(path);

If that doesn't work, then that means your WebAppContext and/or ServletContextHandler does not have a Base Resource directory declared.

> I saw that in Jetty 10 checkGzExists will be removed and not replaced.
> Does this mean the static option using a pre-gzipped file will be removed as well or it will be enabled all the time?
> Is the behavior I am seeing normal, or I did something totally wrong when setting the GzipHandler?

`GzipHandler` doesn't do the static pre-gzipped file serving.  That's the role of `DefaultServlet` (found configured within WebAppContext / ServletContextHandler)
That functionality will still exist, its just that the `checkGzExists` was duplicated effort and pointless to have in `GzipHandler`.
Incidentally, `DefaultServlet` also supports pre-brotli-compressed content as well.

**Comment by lmame** [comment]
Hi @joakime :)
Thanks a lot for the information

For the "jetty.etc.config.urls" and the xmls I will check. Our project started two years ago and in our current jetty /etc folder I think I don't see some of the files you mentioned (I remember a webdefault.xml though but not threadpool for example). I will definitely try what you suggested to figure out the order.

Then I will try again.

**Comment by lmame** [comment]
I looked why we loaded those files and those are the default ones loaded by jetty-osgi:
`public static final String DEFAULT_JETTY_ETC_FILES = "etc/jetty.xml,etc/jetty-http.xml,etc/jetty-deployer.xml";`
 

**Simple test:** 
So I tried to run a very simple test, without osgi, spring, Jax Rs as you suggested (I tested under Windows):
- Got the jetty server 9.4.11.v20180605,
- Created a bundle using the archetype "mvn archetype:generate -DgroupId=com.baeldung -DartifactId=jetty-app -DarchetypeArtifactId=maven-archetype-webapp -DinteractiveMode=false"
- Added in the bundle a js file with flag A, a gz version of it with flag B, modified the index.jsp to load the js script in the <head>,
- Deployed the bundle inside jetty/webapps,

The bundle web.xml is very simple:
```
<!DOCTYPE web-app PUBLIC "-//Sun Microsystems, Inc.//DTD Web Application 2.3//EN" "http://java.sun.com/dtd/web-app_2_3.dtd" >
<web-app>
  <display-name>Archetype Created Web Application</display-name>
</web-app>

```

I launched the jetty using: "java -jar start.jar --add-to-start=gzip", I saw this:
```
INFO  : gzip            initialized in ${jetty.base}\start.ini
INFO  : Base directory was modified
```

I modified the start.ini file to enable "checkGzExists":
`jetty.gzip.checkGzExists=true`

Then I started Jetty with a debug port so I can attach the debugger later:
`java -Xdebug -Xrunjdwp:server=y,transport=dt_socket,address=4000,suspend=n -jar start.jar`

And tried to access the app:
http://localhost:8080/jetty-app/

In chrome debugger I can see that the response header content-encoding is correct (gzip), but the content of the js contains the flag "A".
This means that Jetty server used the file test.js, not the pre-compresse test.js.gz. Then it compressed test.js it and sent it to web browser, gzipped.

I attached the debugger to GzipHandler handle() like before and both the baseRequest.getServletContext() and request.getServletContext() return null.
So the check in "if (_checkGzExists && context!=null)" fails and so there is no check to see if a .gz version of the file exists or not.

Later in the code when the context is set and I can manually check if getRealPath works (evaluate expression). Breakpoint was on _context = context:
org/eclipse/jetty/server/Request.java
```
    public void setContext(Context context)
    {
        _newContext = _context != context;
        _context = context;
    }

```
Evaluation:
`context.getRealPath("/test.js")`
Answer:
`C:\Users\xxxxx\AppData\Local\Temp\jetty-0.0.0.0-8080-jetty-app.war-_jetty-app-any-5588316746096243750.dir\webapp\test.js`


**Back to OSGI:**
The order seemed wrong indeed, here is the order I have when activating gzip with regular 9.4.11 (not osgi):
```
${jetty.base}\etc\jetty-threadpool.xml
${jetty.base}\etc\jetty.xml
${jetty.base}\etc\jetty-webapp.xml
${jetty.base}\etc\jetty-plus.xml
${jetty.base}\etc\jetty-annotations.xml
${jetty.base}\etc\jetty-deploy.xml
${jetty.base}\etc\jetty-gzip.xml
${jetty.base}\etc\home-base-warning.xml
${jetty.base}\etc\jetty-http.xml
```

So I tried this order for OSGI, this does not change anything though, context is still null.
`jetty.etc.config.urls=etc/jetty.xml,etc/jetty-gzip.xml,etc/jetty-http.xml,etc/jetty-deployer.xml`


**In our server code:**
Our code implements Filter and is triggered after the GZipHandler and the Context.
```
import javax.servlet.Filter;
// (...)
public class xxxx implements Filter {
// (...)
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain)
            throws IOException, ServletException {
        HttpServletRequest httpRequest = (HttpServletRequest) request;
        HttpServletResponse httpResponse = (HttpServletResponse) response;
```
At this point the context from "request" is set and I can do things like:
`httpRequest.getServletContext().getMimeType(httpRequest.getPathInfo())`

Anyway I don't know if this is a side effect of OSGI but when I try to get the Real Path of the file the method returns null.
httpRequest.getServletContext().getRealPath(httpRequest.getPathInfo())

**Comment by joakime** [comment]
@janbartel can you take a look at the osgi aspects of this issue?

**Comment by janbartel** [comment]
@lmame  if you use the jetty-gzip.xml file, this will enable gzip for the entire server.  The GzipHandler is called first, before any contexts are consulted, thus the context field is null (because you are not in any context). Pre-existing .gz resources in those contexts are therefore not available to the GzipHandler at this level.

If you want to use precompressed .gz resources in a context, but compress if one is not found, then you should define the GzipHandler to the context in a context xml file.  See the jetty documentation for a description of a context xml file:  http://www.eclipse.org/jetty/documentation/9.4.11.v20180605/configuring-specific-webapp-deployment.html.   For an example of how to configure the GzipHandler for a context in a context xml file, look at: https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/tests/test-webapps/test-jetty-webapp/src/main/config/demo-base/webapps/test.xml.  Take a look at the setters on GzipHandler to know what you can configure on it in a context xml file: http://www.eclipse.org/jetty/javadoc/9.4.11.v20180605/org/eclipse/jetty/server/handler/gzip/GzipHandler.html

See here for info on how to provide a context xml file in a webapp bundle: http://www.eclipse.org/jetty/documentation/9.4.11.v20180605/framework-jetty-osgi.html#_deploying_bundles_as_webapps

**Comment by janbartel** [comment]
@lmame  I should also have added that in order to get the DefaultServlet to use the precompressed files, you'll also need to set up an init-param on it. One way to do that is to add the following lines to your web.xml file:
```xml
  <servlet>
     <servlet-name>default</servlet-name>
     <init-param>
        <param-name>precompressed</param-name>
        <param-value>true</param-value>
    </init-param>
    </servlet>
```

Otherwise, as you have found, the GzipHandler will compress the plain file on-the-fly.

**Comment by janbartel** [comment]
I've opened issues #2844 and #2845 to clean up our documentation on gzipping and the DefaultServlet.

**Comment by lmame** [comment]
Hi @janbartel , thanks a lot, I will give it a try this week end :)

**Comment by lmame** [comment]
Hi @janbartel, before testing with OSGI I tried to apply it to my simple example and, well, I might be doing something wrong because I can't make it work :)

So I have a A.js file and its gz versions A.js.gz (and A.js.svgz).
The gz versions are a bit different so I know if Jetty serves the already gzipped versions or gzips A.js.


**Simple tests using Jetty server 9.4.11.v20180605 (OOTB, no OSGI):** 
_Test 1, no global gzip set and no bundle context gzip set:_
- I do not activate gzip globally in start.ini,
- I use the example async-rest.war from /demo-base/webapps,
- I modify the example war directly:
- index.html page example to load a js (A.js) in the head (<script type="text/javascript" src="A.js"></script>),
- added files A.js (it is 1.8 MB) and A.js.gz, A.js.svgz in the war (same level as A.js),
- Put the war inside the Jetty /webapps folder,
- Started the server (java -Xdebug -Xrunjdwp:server=y,transport=dt_socket,address=4000,suspend=n -jar start.jar),
- Opened the page http://localhost:8080/async-rest/
- In chrome debugger network tab I see the get on the A.js comes from Jetty not gzipped which is normal since gzip is not enabled,


_Test 2, no global gzip set and bundle context gzip enabled:_
- I create a specific context file for async-rest in /webapps/async-rest.xml enabling gzip for this bundle:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "http://www.eclipse.org/jetty/configure_9_3.dtd">
<Configure class="org.eclipse.jetty.webapp.WebAppContext">
    <Set name="contextPath">/async-rest</Set>
    <Set name="war"><Property name="jetty.webapps"/>/async-rest.war</Set>
    <Set name="gzipHandler">
        <New class="org.eclipse.jetty.server.handler.gzip.GzipHandler">
            <Set name="minGzipSize">2048</Set>
        </New>
    </Set>
</Configure>
```
- restart Jetty server,
- Opened the page http://localhost:8080/async-rest/
- In chrome debugger network tab I see the get on the A.js comes from Jetty and is gzipped with the content of A.js, so ok,


_Test 3, no global gzip set, bundle context gzip enabled and precompressed gzip enabled in bundle:_
- I modify async-rest/WEB-INF/web.xml to add the init-param for the default servlet to enable the use of precompressed files:
```
<?xml version="1.0"?>
<web-app xmlns="http://xmlns.jcp.org/xml/ns/javaee" 
   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
   xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_3_1.xsd" 
   version="3.1">
	<display-name>Async REST Webservice Example</display-name>
	  <servlet>
     <servlet-name>default</servlet-name>
     <init-param>
        <param-name>precompressed</param-name>
        <param-value>true</param-value>
    </init-param>
    </servlet>
</web-app>
```
- restart Jetty server,
- Opened the page http://localhost:8080/async-rest/
- In chrome debugger network tab I see the get on the A.js comes from Jetty and is gzipped, so ok, but the content is still the one from A.js so Jetty did not use precompressed files,

**Comment by lmame** [comment]
Bleh...
Looking in the code there is a condition that explains why it didn't work for my tests:
if (compressedResource != null && compressedResource.exists() && compressedResource.lastModified() >= resource.lastModified()
                        && compressedResource.length() < resource.length())

Especially this one, my A.js was more recent than the A,js.gz:
compressedResource.lastModified() >= resource.lastModified()

Test 3 above works now, I'll try to enable global gzip and precompressed in bundle.

**Comment by lmame** [comment]
**Note:**
I removed one message that was confusing as it was a failed test due to the file dates I mention above.
I summarize my tests below.


Hi @janbartel, before testing with OSGI I tried to apply it to my simple example and it seems to be working :)

So I have a A.js file and its gz versions A.js.gz (and A.js.svgz).
The gz versions are a bit different so I know if Jetty serves the already gzipped versions or gzips A.js.


**Simple tests using Jetty server 9.4.11.v20180605 (OOTB, no OSGI):** 
_Test 1, no global gzip set and no bundle context gzip set:_
- I do not activate gzip globally in start.ini,
- I use the example async-rest.war from /demo-base/webapps,
- I modify the example war directly:
- index.html page example to load a js (A.js) in the head (<script type="text/javascript" src="A.js"></script>),
- added files A.js (it is 1.8 MB) and A.js.gz, A.js.svgz in the war (same level as A.js),
- Put the war inside the Jetty /webapps folder,
- Started the server (java -Xdebug -Xrunjdwp:server=y,transport=dt_socket,address=4000,suspend=n -jar start.jar),
- Opened the page http://localhost:8080/async-rest/
- In chrome debugger network tab I see the get on the A.js comes from Jetty not gzipped which is normal since gzip is not enabled,


_Test 2, no global gzip set and bundle context gzip enabled:_
- I create a specific context file for async-rest in /webapps/async-rest.xml enabling gzip for this bundle:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "http://www.eclipse.org/jetty/configure_9_3.dtd">
<Configure class="org.eclipse.jetty.webapp.WebAppContext">
    <Set name="contextPath">/async-rest</Set>
    <Set name="war"><Property name="jetty.webapps"/>/async-rest.war</Set>
    <Set name="gzipHandler">
        <New class="org.eclipse.jetty.server.handler.gzip.GzipHandler">
            <Set name="minGzipSize">2048</Set>
        </New>
    </Set>
</Configure>
```
- restart Jetty server,
- Opened the page http://localhost:8080/async-rest/
- In chrome debugger network tab I see the get on the A.js comes from Jetty and is gzipped with the content of A.js, so ok,


_Test 3, no global gzip set, bundle context gzip enabled and precompressed gzip enabled in bundle:_
- I modify async-rest/WEB-INF/web.xml to add the init-param for the default servlet to enable the use of precompressed files:
```
<?xml version="1.0"?>
<web-app xmlns="http://xmlns.jcp.org/xml/ns/javaee" 
   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
   xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_3_1.xsd" 
   version="3.1">
	<display-name>Async REST Webservice Example</display-name>
	  <servlet>
     <servlet-name>default</servlet-name>
     <init-param>
        <param-name>precompressed</param-name>
        <param-value>true</param-value>
    </init-param>
    </servlet>
</web-app>
```
- restart Jetty server,
- Opened the page http://localhost:8080/async-rest/
- In chrome debugger network tab I see the get on the A.js comes from Jetty and is gzipped and the content is the one of A.js.gz so correct :)


_Test 4, no global gzip set, no bundle context and precompressed gzip enabled in bundle:_
- I still have the modified async-rest/WEB-INF/web.xml from step 3 that has the init-param precompressed to true,
- I removed the /webapps/aysnc-rest.xml context file,
- restart Jetty server,
- Opened the page http://localhost:8080/async-rest/
- In chrome debugger network tab I see the get on the A.js comes from Jetty and is gzipped and the content is the one of A.js.gz so correct :)


_Test 5, global gzip set, no bundle context and precompressed gzip enabled in bundle:_
- I still have the modified async-rest/WEB-INF/web.xml from step 3 that has the init-param precompressed to true,
- I don't have a context file for the bundle,
- I have enabled gzip compression at Jetty level in start.ini:
```
--module=gzip
jetty.gzip.minGzipSize=2048

```
- restart Jetty server,
- Opened the page http://localhost:8080/async-rest/
- In chrome debugger network tab I see the get on the A.js comes from Jetty and is gzipped and the content is the one of A.js.gz so correct :)
- If in the bundle I remove the A.js.gz and I redeploy the bundle then I can see in Chrome debugger that the file comes gzipped (by Jetty) so it's correct,


To summarize:

- Global gzip Off + bundle (with A.js and A.js.gz): No compression,
- Global gzip On (start.ini) + bundle (with A.js and A.js.gz): Compressed by Jetty,
- Global gzip Off + context bundle with Gzip On + bundle (with A.js and A.js.gz): Compressed by Jetty,
- Global gzip Off + context bundle with Gzip On + bundle (with A.js and A.js.gz) with precompressed True: returns A.js.gz,
- Global gzip Off + bundle (with A.js and A.js.gz, B,js) with precompressed True: returns A.js.gz, B.js comes compressed by Jetty,


So it seems it works, at least with OOTB Jetty I'll run the tests with OSGI. Thanks a lot for the help :)

**Comment by lmame** [comment]
I updated our server to use 9.4.11.v20180605 since our 9.3.7.v20160115 did not have the precompressed code.
I have two bundles where I added some .gz files and the web.xml.
When our server starts I've set a breakoint in "org/eclipse/jetty/servlet/DefaultServlet.java" in "private CompressedContentFormat[] parsePrecompressedFormats(String precompressed, boolean gzip)".
I can see for both those bundles that the web.xml is read correctly with precompressed=true and it returns ret as array of 2 (br/gzip).

Though once the server is fully started when I load a page, the precompressed .gz is not used (the file  timestamp is correct now :)) and when the code stops in "org/eclipse/jetty/server/handler/gzip/GzipHandler.java" "public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException" the context is null as before.

In the OOTB server the breakpoint I've set in "org/eclipse/jetty/server/ResourceContentFactory.java" "private HttpContent load(String pathInContext, Resource resource, int maxBufferSize)" was hit.
Not in my OSGI server.

**Comment by janbartel** [comment]
@lmame I just tested with 9.4.12 and osgi and it worked fine. In fact, I didn't configure the GzipHandler at all, just configured precompressed content for the DefaultServlet, so the fact that I got compressed content back is entirely due to the behaviour of DefaultServlet/ResourceService.

Can you enable DEBUG level for org.eclipse.jetty.server.ResourceService and test retrieving the static file  again?

**Comment by janbartel** [comment]
@lmame ?

**Comment by lmame** [comment]
Hi @janbartel , I am really sorry for the delay...

I am working overtime / overdrive / swamped for my company right now, I will try to send the logs during the week, this week end worst case scenario.

I will update our server to use the version you tested (9.4.12), enable the logs, configure precompressed for the DefaultServlet and test again.

**Comment by joakime** [comment]
No reproduction provided. Closing.
